contracts/extensions/Tasks.sol

```diff
@@ -0,0 +1,677 @@
+  // https://github.com/ethereum/solidity/issues/2884
+  function executeCall(address to, uint256 value, bytes memory data) internal returns (bool success) {
+    assembly {
+      success := call(gas, to, value, add(data, 0x20), mload(data), 0, 0)
+    }
+  }
+
+  // Get the function signature and task id from the transaction bytes data
+  // Note: Relies on the encoded function's first parameter to be the uint256 taskId
+  function deconstructCall(bytes memory _data) internal pure returns (bytes4 sig, uint256 taskId) {
+    assembly {
+      sig := mload(add(_data, 0x20))
+      taskId := mload(add(_data, 0x24)) // same as calldataload(72)
+    }
+  }
+
+  function deconstructRoleChangeCall(bytes memory _data) internal pure returns (bytes4 sig, uint256 taskId, address userAddress) {
+    assembly {
+      sig := mload(add(_data, 0x20))
+      taskId := mload(add(_data, 0x24)) // same as calldataload(72)
+      userAddress := mload(add(_data, 0x44))
+    }
+  }
+
+  function taskWorkRatingsAssigned(uint256 _id) internal view returns (bool) {
+    Role storage workerRole = taskRoles[_id][uint8(TaskRole.Worker)];
+    Role storage managerRole = taskRoles[_id][uint8(TaskRole.Manager)];
+    return (workerRole.rating != TaskRatings.None) && (managerRole.rating != TaskRatings.None);
+  }
+
+  function taskWorkRatingsClosed(uint256 _id) internal view returns (bool) {
+    assert(tasks[_id].completionTimestamp > 0);
+    assert(ratingSecrets[_id].count <= 2);
+    if (ratingSecrets[_id].count == 2) {
+      return sub(now, ratingSecrets[_id].timestamp) > RATING_REVEAL_TIMEOUT;
+    } else {
+      return sub(now, tasks[_id].completionTimestamp) > add(RATING_COMMIT_TIMEOUT, RATING_REVEAL_TIMEOUT);
+    }
+  }
+
+  function assignWorkRatings(uint256 _id) internal {
+    require(taskWorkRatingsAssigned(_id) || taskWorkRatingsClosed(_id), "colony-task-ratings-not-closed");
+
+    // In the event of a user not committing/revealing within the rating window,
+    // their rating of their counterpart is assumed to be the maximum
+    // and they will receive a (payout/2) reputation penalty
+
+    Role storage managerRole = taskRoles[_id][uint8(TaskRole.Manager)];
+    Role storage workerRole = taskRoles[_id][uint8(TaskRole.Worker)];
+    Role storage evaluatorRole = taskRoles[_id][uint8(TaskRole.Evaluator)];
+
+    if (workerRole.rating == TaskRatings.None) {
+      evaluatorRole.rating = TaskRatings.Unsatisfactory;
+      evaluatorRole.rateFail = true;
+      workerRole.rating = TaskRatings.Excellent;
+    } else {
+      evaluatorRole.rating = TaskRatings.Satisfactory;
+    }
+
+    if (managerRole.rating == TaskRatings.None) {
+      workerRole.rateFail = true;
+      managerRole.rating = TaskRatings.Excellent;
+    }
+  }
+
+  function setTaskRoleUser(uint256 _id, TaskRole _role, address payable _user)
+    private
+    taskExists(_id)
+    taskNotComplete(_id)
+  {
+    taskRoles[_id][uint8(_role)] = Role({
+      user: _user,
```

Comment:
Why not use `recipient` in the corresponding expenditure slot rather than storing the address again here?

Reply:
No reason not too -- it crossed my mind towards the end but in the moment I just didn't feel like adding all of the `colony.getExpenditureRecipient` queries. But I can.

Reply:
I would vastly prefer this - crucially, it would mean that the roles could be disputed.

Reply:
It will be done!